sw/host/opentitanlib/src/util/usb.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{ensure, Result};
+use rusb;
+use std::time::Duration;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Could not find USB device")]
+    NotFound,
+    #[error("Found multiple USB devices, use --serial")]
+    MultipleDevices,
+}
+
+/// The `UsbBackend` provides low-level USB access to debugging devices.
+pub struct UsbBackend {
+    device: rusb::Device<rusb::GlobalContext>,
+    handle: rusb::DeviceHandle<rusb::GlobalContext>,
+    serial_number: String,
+    timeout: Duration,
+}
+
+impl UsbBackend {
+    /// Scan the USB bus for a device matching VID/PID, and optionally also matching a serial
+    /// number.
+    pub fn scan(
+        usb_vid: u16,
+        usb_pid: u16,
+        usb_serial: &Option<String>,
+    ) -> Result<Vec<(rusb::Device<rusb::GlobalContext>, String)>> {
+        let mut devices = Vec::new();
+        for device in rusb::devices()?.iter() {
+            let descriptor = match device.device_descriptor() {
+                Ok(desc) => desc,
+                _ => {
+                    log::error!(
+                        "Could not read device descriptor for device at bus={} address={}",
+                        device.bus_number(),
+                        device.address()
+                    );
+                    continue;
+                }
+            };
+            if descriptor.vendor_id() != usb_vid {
+                continue;
+            }
+            if descriptor.product_id() != usb_pid {
+                continue;
+            }
+            let handle = match device.open() {
+                Ok(handle) => handle,
+                _ => {
+                    log::error!(
+                        "Could not open device at bus={} address={}",
+                        device.bus_number(),
+                        device.address()
+                    );
+                    continue;
+                }
+            };
+            let serial_number = match handle.read_serial_number_string_ascii(&descriptor) {
+                Ok(sn) => sn,
+                _ => {
+                    log::error!(
+                        "Could not read serial number from device at bus={} address={}",
+                        device.bus_number(),
+                        device.address()
+                    );
+                    continue;
+                }
+            };
+            if let Some(sn) = &usb_serial {
+                if &serial_number != sn {
+                    continue;
+                }
+            }
+            devices.push((device, serial_number));
+        }
+        Ok(devices)
+    }
+
+    /// Create a new UsbBackend.
+    pub fn new(
+        usb_vid: u16,
+        usb_pid: u16,
+        usb_serial: &Option<String>,
```

Comment:
Wouldn't `Option<&str>` make more sense?

Reply:
I believe it does. Done.